query/src/main/java/io/keen/client/java/Funnel.java

```diff
@@ -0,0 +1,214 @@
+        // Validate step properties that cannot be true for the first funnel step
+        FunnelStep firstStep = builder.steps.get(0);
+        
+        if (null != firstStep.getInverted() &&
+            true == firstStep.getInverted()) {
+            throw new IllegalArgumentException(
+                "First step in funnel cannot have special parameter 'inverted' set to true.");
+        }
+        
+        if (null != firstStep.getOptional() &&
+            true == firstStep.getOptional()) {
+            throw new IllegalArgumentException(
+                "First step in funnel cannot have special parameter 'optional' set to true.");
+        }
+        
+        // The steps in the Builder are ordered, so the request params will be too.
+        this.steps = new RequestParameterCollection<FunnelStep>(builder.steps);
+    }
+    
+    /**
+     * Get the URL for this request.
+     * 
+     * @param urlBuilder The RequestUrlBuilder instance to use for building the URL.
+     * @param projectId The projectId to use for the URL.
+     * @return The URL for the request.
+     * @throws KeenQueryClientException Thrown if there are errors formatting the URL.
+     */
+    @Override
+    URL getRequestURL(RequestUrlBuilder urlBuilder, String projectId) throws KeenQueryClientException {
+        return urlBuilder.getAnalysisUrl(projectId, KeenQueryConstants.FUNNEL);
+    }
+
+    /**
+     * Construct the jsonifiable arguments for this request.
+     * 
+     * @return A jsonifiable Map to use for the request body.
+     */
+    @Override
+    Map<String, Object> constructRequestArgs() {
+        Map<String, Object> args = new HashMap<String, Object>();
+        
+        args.put(KeenQueryConstants.STEPS, this.steps.constructParameterRequestArgs());
+        
+        if (null != this.timeframe) {
+            args.putAll(timeframe.constructTimeframeArgs());
+        }
+        
+        return args;
+    }
+
+    @Override
+    boolean groupedResponseExpected() { return false; }
+
+    @Override
+    boolean intervalResponseExpected() { return false; }
+    
+    /**
+     * Builder for creating a Funnel query.
+     */
+    public static class Builder {
+        private List<FunnelStep> steps;
+        private Timeframe timeframe;
+        
+        /**
+         * Set the list of funnel steps. Existing steps will be lost.
+         * 
+         * @param steps The funnel steps.
+         */
+        public void setSteps(List<? extends FunnelStep> steps) {
+            this.steps = null;
+
+            // Client code may just be clearing all the steps.
+            if (null != steps) {
+                withSteps(steps); // Shallow copy the list of steps
+            }
+        }
+        
+        /**
+         * Set the list of funnel steps.
+         * 
+         * @param steps The funnel steps.
+         * @return This Builder instance with the steps added.
+         */
+        public Builder withSteps(List<? extends FunnelStep> steps) {
```

Comment:
We are going to spin up a v6 branch internally and toss this sort of thing into it, then. Thanks!